### PR TITLE
`Development`: Refactor participation and submission services

### PR DESCRIPTION
**Pull request by FlorianGlombik**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Fixes https://github.com/ls1intum/Artemis/issues/4464 and resolves https://github.com/ls1intum/Artemis/issues/4428
In the refactoring [General: Improve access right management in client services #3742](https://github.com/ls1intum/Artemis/pull/3742) I have missed refactoring the participation components and services.
This lead to the issue https://github.com/ls1intum/Artemis/issues/4464 and should be refactored to increase the consistency and code quality of Artemis.

### Description

- Removed additional `isAtLeast...` attributes, as they are redundant with values that are already stored in entities as `courses`  and `exercises`
- Refactored services to remove code duplicates and make modifications easier for changes in the future
- Refactored the corresponding `html` files accordingly _(and using `ng-else` for the `exam detail` view)_
- Added the `Archive Exam` functionality to the exam functionality overview in the documentation

### Steps for Testing

_This PR refactored many services, so potentially the functionality of all exercise-related steps are affected (even creating exercises, submitting solutions - if invalid objects are sent to the server). Nothing should have changed here, but you might want to play a bit around with creating and especially participating in exercises, however, issues are probably rather found in code reviews here._

Prerequisites:
- 1 Instructor
- 2 Tutors
- 1 exercise of each type

#### Create a submission for the exercises
1. Make sure you can participate in exercises _(participate as Student/Tutor/Instructor, this should not matter)_

#### Assess the Submissions
2. Assess the submissions with the `tutor`
3. `Instructors` should be able to **overwrite locked assessments** - meaning that the buttons are displayed and work as expected

#### Assess complaints
4. Complaint with the user you made the submission in step1
5. Answer the complaint with another `tutor`
6. `Instructors` should still be able to **overwrite the assessment**

#### Further Steps

7. Make sure the `exam detail` view is not broken for `instructors`
8. Verify that the access rights documentation is not broken due to the addition of `Archive Exam` in the access matrix _(you can find the current documentation in the GitHub actions if you do not want to build it locally)_
<details>
<summary>Documentation GitHub action</summary>

image
</details>

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

<details>
<summary>Instructors should be able to overwrite locked assessments (programming / file upload)</summary>
The buttons should be visible and work as expected.

image

So it should not look like this for instructors, as it was the case before:
image</details>

<details>
<summary>Instructors should be able to overwrite locked assessments (text / modeling)</summary>
The buttons should be visible and work as expected.

image

So it should not look like this for instructors, as it was the case before:
image</details>

<details>
<summary>Exam detail view</summary>

Opening the exam detail view as `instructor` shows the functionality available for exams and the `exam checklist` first. When scrolling down you can find some information about the exam - this is where I made some refactorings. 

<details>
<summary>Exam detail view without scrolling down</summary>

image
</details>

If the values are displayed properly and for not present values `Not set` is displayed the refactoring did not break something here.

image
</details>

<details>
<summary>Updated documentation</summary>

image</details>

### CODE REVIEW

**Review by JohannesStoehr — APPROVED** [review]
Looked through the code and did not find any flaws. I only marked two places where I think the code could be improved a bit more

File: src/main/webapp/app/exam/manage/exams/exam-detail.component.html [hunk 1]
```diff
@@ -8,56 +8,67 @@ <h2>
             <hr />
             <jhi-alert></jhi-alert>
             <jhi-alert-error></jhi-alert-error>
-            <div>
+            <div *ngIf="exam.course">
```
[the remark this answers is not in the document]
Comment (FlorianGlombik) [reply src/main/webapp/app/exam/manage/exams/exam-detail.component.html]: Thanks for the heads up, the `archive course` button does actually require instructor rights to be executed _(see `archiveExam` in `ExamResource.java`)_.
I moved the check to the suggested `div`. 👍

[the remark this answers is not in the document]
Comment (FlorianGlombik) [reply src/main/webapp/app/exam/manage/exams/exam-detail.component.html]: Addressed in https://github.com/ls1intum/Artemis/pull/4468/commits/07443d81d5a14689f6e9b7a2b575b214f4b43077 and https://github.com/ls1intum/Artemis/pull/4468/commits/ba230405b4d954f7262a23c8b3ebf31a3db34b96 - when Looking at the component I noticed redundant checks for displaying exam information and exchanged them with else statements right away https://github.com/ls1intum/Artemis/pull/4468/commits/d974d2adef0008dbfa6a69f058ab3d7a27bc77f9

File: src/main/webapp/app/exam/manage/exams/exam-checklist-component/exam-checklist.component.html [hunk 2]
```diff
@@ -88,7 +88,7 @@ <h3>
                 </td>
                 <td>
                     <a
-                        *ngIf="isAtLeastInstructor"
+                        *ngIf="exam!.course!.isAtLeastInstructor"
```
[the remark this answers is not in the document]
Comment (FlorianGlombik) [reply src/main/webapp/app/exam/manage/exams/exam-checklist-component/exam-checklist.component.html]: Yes, it does - fixed in https://github.com/ls1intum/Artemis/pull/4468/commits/975208bb1b358b1db27828ef90b14fcab5e8c547

File: src/main/webapp/app/exercises/programming/manage/services/programming-exercise.service.ts [hunk 3]
```diff
@@ -225,25 +207,19 @@ export class ProgrammingExerciseService {
...
-            );
+        return this.http.get<ProgrammingExercise>(`${this.resourceUrl}/${programmingExerciseId}/with-template-and-solution-participation`, { params, observe: 'response' }).pipe(
+            map((res: EntityResponseType) => {
+                if (res.body && withSubmissionResults) {
+                    // We need to reconnect the submissions with the results. They got removed because of the circular dependency
+                    const templateSubmissions = res.body.templateParticipation?.submissions;
+                    this.reconnectSubmissionAndResult(templateSubmissions);
+                    const solutionSubmissions = res.body.solutionParticipation?.submissions;
+                    this.reconnectSubmissionAndResult(solutionSubmissions);
+                    this.processProgrammingExerciseEntityResponse(res);
```
Comment (b-fein) [inline src/main/webapp/app/exercises/programming/manage/services/programming-exercise.service.ts:218]:
```suggestion
                    this.reconnectSubmissionAndResult(solutionSubmissions);

                    this.processProgrammingExerciseEntityResponse(res);
```
I wondered why the conversion is no longer needed before I spotted this method call. The additional newline could maybe increase its visibility.

**Review by b-fein — APPROVED** [review]
Code looks good, I found only two minor possible improvements.

File: src/main/webapp/app/exercises/text/assess/text-assessment.service.ts [hunk 4]
```diff
@@ -213,17 +219,22 @@ export class TextAssessmentService {
...
-        if (result.submission && result.submission.submissionDate) {
+        if (result.submission?.submissionDate) {
             result.submission.submissionDate = dayjs(result.submission.submissionDate);
         }
-        if (result.participation && result.participation.initializationDate) {
-            result.participation.initializationDate = dayjs(result.participation.initializationDate);
+        if (result.participation) {
+            if (result.participation.initializationDate) {
+                result.participation.initializationDate = dayjs(result.participation.initializationDate);
+            }
```
[the remark this answers is not in the document]
Comment (FlorianGlombik) [reply src/main/webapp/app/exercises/text/assess/text-assessment.service.ts:234]: Thanks for the heads up, I have missed that 👍

**Comment by FlorianGlombik** [comment]
Thank you for the review and the suggested changes @b-fein!

**Review by VinnavanSembian — DISMISSED** [review]
tested on ts1 with @lBruchmann @Strohgelaender @willergeorg. Everything works as expected!

**Review by lBruchmann — APPROVED** [review]
Tested on ts1. Tested together with @Strohgelaender, @VinnavanSembian, and @willergeorg. Submission services work as expected. Tested for each exercise type.

**Review by willergeorg — DISMISSED** [review]
Tested on TS1 during testing session. We tested every exercise type except quizzes and followed your testing steps. No errors visible.

**Review by Strohgelaender — DISMISSED** [review]
Tested on TS1 @lBruchmann @VinnavanSembian @willergeorg. Creating and participating in exercises worked and after the deadline tutors were able to provide feedback. Instructors were able to overwrite feedback at any stage (before the assessment due date, before submitting a complaint, after submitting a complaint, and after answering the complaint). I also looked at the exam detail view, I did not see any problems here.

**Review by baranimo7 — APPROVED** [review]
Followed the testing steps and everything works as expected

**Review by MertAksehirlioglu — APPROVED** [review]
Tested on TS2, works as expected.

**Pull request merged**